File: server/src/models/hr.rs
use async_graphql::Enum;
use mongodb::bson::{oid::ObjectId, DateTime};
use serde::{Deserialize, Serialize};

// ============================================================================
// ENUMS
// ============================================================================

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Enum)]
pub enum LeaveTypeEnum {
    #[graphql(name = "Annual")]
    Annual,
    #[graphql(name = "Sick")]
    Sick,
    #[graphql(name = "Maternity")]
    Maternity,
    #[graphql(name = "Personal")]
    Personal,
    #[graphql(name = "Unpaid")]
    Unpaid,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Enum)]
pub enum LeaveStatus {
    #[graphql(name = "Pending")]
    Pending,
    #[graphql(name = "Approved")]
    Approved,
    #[graphql(name = "Rejected")]
    Rejected,
}

// ============================================================================
// STAFF
// ============================================================================

#[derive(Debug, Serialize, Deserialize)]
pub struct Staff {
    #[serde(rename = "_id", skip_serializing_if = "Option::is_none")]
    pub id: Option<ObjectId>,
    pub school_id: String,
    pub staff_id: String,
    pub first_name: String,
    pub last_name: String,
    pub email: String,
    pub phone: String,
    pub date_of_birth: String,
    pub gender: String,
    pub address: String,
    pub role: String, // "teacher" | "admin" | "principal" | "librarian"
    pub department: Option<String>,
    pub subjects: Vec<String>,
    pub hire_date: DateTime,
    pub salary: f64,
    pub currency: String,
    pub status: String, // "active" | "inactive" | "on_leave"
    pub profile_photo: Option<String>,
    pub created_at: DateTime,
    pub updated_at: DateTime,
}

// ============================================================================
// PAYROLL
// ============================================================================

#[derive(Debug, Serialize, Deserialize)]
pub struct Payroll {
    #[serde(rename = "_id", skip_serializing_if = "Option::is_none")]
    pub id: Option<ObjectId>,
    pub school_id: String,
    pub staff_id: ObjectId,
    pub staff_name: Option<String>,
    pub month: String, // "2024-01"
    pub base_salary: f64,
    pub bonuses: f64,
    pub deductions: f64,
    pub net_salary: f64,
    pub currency: String,
    pub payment_date: Option<DateTime>,
    pub status: String, // "pending" | "paid"
    pub created_at: DateTime,
    pub updated_at: DateTime,
}

// ============================================================================
// LEAVE REQUEST
// ============================================================================

#[derive(Debug, Serialize, Deserialize)]
pub struct Leave {
    #[serde(rename = "_id", skip_serializing_if = "Option::is_none")]
    pub id: Option<ObjectId>,
    pub school_id: String,
    pub staff_id: ObjectId,
    pub leave_type: LeaveTypeEnum,
    pub start_date: String,
    pub end_date: String,
    pub days: i32,
    pub reason: String,
    pub status: LeaveStatus,
    pub approved_by: Option<String>,
    pub rejection_reason: Option<String>,
    pub created_at: DateTime,
    pub updated_at: DateTime,
}

// ============================================================================
// LEAVE BALANCE
// ============================================================================

#[derive(Debug, Serialize, Deserialize)]
pub struct LeaveBalance {
    #[serde(rename = "_id", skip_serializing_if = "Option::is_none")]
    pub id: Option<ObjectId>,
    pub school_id: String,
    pub staff_id: ObjectId,
    pub academic_year: String, // "2025-2026"
    pub annual_total: i32,
    pub annual_used: i32,
    pub sick_total: i32,
    pub sick_used: i32,
    pub maternity_total: i32,
    pub maternity_used: i32,
    pub personal_total: i32,
    pub personal_used: i32,
    pub created_at: DateTime,
    pub updated_at: DateTime,
}

File: dashboard/app/graphql/hr.ts
// HR GraphQL Queries and Mutations

const STAFF_FIELDS = `
	id
	schoolId
	staffId
	firstName
	lastName
	email
	phone
	dateOfBirth
	gender
	address
	role
	department
	subjects
	hireDate
	salary
	currency
	status
	profilePhoto
	createdAt
	updatedAt
`;

const PAYROLL_FIELDS = `
	id
	schoolId
	staffId
	staffName
	month
	baseSalary
	bonuses
	deductions
	netSalary
	currency
	paymentDate
	status
	createdAt
	updatedAt
`;

const LEAVE_FIELDS = `
	id
	schoolId
	staffId
	leaveType
	startDate
	endDate
	days
	reason
	status
	approvedBy
	rejectionReason
	createdAt
	updatedAt
`;

const LEAVE_BALANCE_FIELDS = `
	id
	schoolId
	staffId
	academicYear
	annualTotal
	annualUsed
	annualRemaining
	sickTotal
	sickUsed
	sickRemaining
	maternityTotal
	maternityUsed
	maternityRemaining
	personalTotal
	personalUsed
	personalRemaining
`;

export const HR_QUERIES = {
	ALL_STAFF: `
		query AllStaff($schoolId: String!) {
			allStaff(schoolId: $schoolId) {
				${STAFF_FIELDS}
			}
		}
	`,

	STAFF_BY_ID: `
		query Staff($id: String!) {
			staff(id: $id) {
				${STAFF_FIELDS}
			}
		}
	`,

	PAYROLL_FOR_STAFF: `
		query PayrollForStaff($staffId: String!) {
			payrollForStaff(staffId: $staffId) {
				${PAYROLL_FIELDS}
			}
		}
	`,

	PAYROLL_HISTORY: `
		query PayrollHistory($schoolId: String!, $month: String, $page: Int, $pageSize: Int) {
			payrollHistory(schoolId: $schoolId, month: $month, page: $page, pageSize: $pageSize) {
				items {
					${PAYROLL_FIELDS}
				}
				total
				page
				totalPages
			}
		}
	`,

	STAFF_LEAVES: `
		query StaffLeaves($schoolId: String!, $staffId: String, $status: String) {
			staffLeaves(schoolId: $schoolId, staffId: $staffId, status: $status) {
				${LEAVE_FIELDS}
			}
		}
	`,

	LEAVE_BALANCE: `
		query LeaveBalance($schoolId: String!, $staffId: String!, $academicYear: String!) {
			leaveBalance(schoolId: $schoolId, staffId: $staffId, academicYear: $academicYear) {
				${LEAVE_BALANCE_FIELDS}
			}
		}
	`,

	ALL_LEAVE_BALANCES: `
		query AllLeaveBalances($schoolId: String!, $academicYear: String!) {
			allLeaveBalances(schoolId: $schoolId, academicYear: $academicYear) {
				${LEAVE_BALANCE_FIELDS}
			}
		}
	`,
};

export const HR_MUTATIONS = {
	CREATE_STAFF: `
		mutation CreateStaff($input: CreateStaffInput!) {
			createStaff(input: $input) {
				${STAFF_FIELDS}
			}
		}
	`,

	UPDATE_STAFF: `
		mutation UpdateStaff($id: String!, $input: UpdateStaffInput!) {
			updateStaff(id: $id, input: $input) {
				${STAFF_FIELDS}
			}
		}
	`,

	DELETE_STAFF: `
		mutation DeleteStaff($id: String!) {
			deleteStaff(id: $id)
		}
	`,

	CREATE_PAYROLL: `
		mutation CreatePayroll($input: CreatePayrollInput!) {
			createPayroll(input: $input) {
				${PAYROLL_FIELDS}
			}
		}
	`,

	GENERATE_MONTHLY_PAYROLL: `
		mutation GenerateMonthlyPayroll($input: GeneratePayrollInput!) {
			generateMonthlyPayroll(input: $input) {
				generatedCount
				totalAmount
				month
				payrolls {
					${PAYROLL_FIELDS}
				}
			}
		}
	`,

	UPDATE_PAYROLL_STATUS: `
		mutation UpdatePayrollStatus($id: String!, $input: UpdatePayrollStatusInput!) {
			updatePayrollStatus(id: $id, input: $input) {
				${PAYROLL_FIELDS}
			}
		}
	`,

	CREATE_LEAVE: `
		mutation CreateLeave($input: CreateLeaveInput!) {
			createLeave(input: $input) {
				${LEAVE_FIELDS}
			}
		}
	`,

	UPDATE_LEAVE_STATUS: `
		mutation UpdateLeaveStatus($id: String!, $input: UpdateLeaveStatusInput!) {
			updateLeaveStatus(id: $id, input: $input) {
				${LEAVE_FIELDS}
			}
		}
	`,

	INIT_LEAVE_BALANCE: `
		mutation InitLeaveBalance($input: InitLeaveBalanceInput!) {
			initLeaveBalance(input: $input) {
				${LEAVE_BALANCE_FIELDS}
			}
		}
	`,
};
